File: 502-ipo/ipo.py

class Project:
    def __init__(self, profit, capital):
        self.profit = profit
        self.capital = capital

class Solution(object):
    def findMaximizedCapital(self, k, w, profits, capital):
        """
        :type k: int
        :type w: int
        :type profits: List[int]
        :type capital: List[int]
        :rtype: int
        """
        n = len(profits)
        projects = [Project(profits[i], capital[i]) for i in range(n)]

        projects.sort(key=lambda x: x.capital)

        i = 0

        maximizeCapital = []
        while k > 0:
            while i < n and projects[i].capital <= w:
                heapq.heappush(maximizeCapital, (-projects[i].profit, projects[i].capital))
                i += 1
            if not maximizeCapital:
                break
            profit, project_capital = heapq.heappop(maximizeCapital)
            w += -profit
            k -= 1

        return w

File: 502-ipo/ipo.java
class Project {
    int profit;
    int capital;

    public Project(int profit, int capital) {
        this.profit = profit;
        this.capital = capital;
    }
}

class Solution {
    public int findMaximizedCapital(int k, int w, int[] profits, int[] capital) {
        int n = profits.length;
        Project[] projects = new Project[n];

        for (int i = 0; i < n; i++) {
            projects[i] = new Project(profits[i], capital[i]);
        }

        Arrays.sort(projects, Comparator.comparingInt(a -> a.capital));

        int i = 0;

        PriorityQueue<Project> maximizeCapital = new PriorityQueue<>((a, b) -> Integer.compare(b.profit, a.profit));

        while (k-- > 0) {
            while (i < n && projects[i].capital <= w) {
                maximizeCapital.offer(projects[i]);
                i++;
            }
            if (maximizeCapital.isEmpty()) {
                break;
            }
            w += maximizeCapital.poll().profit;
        }

        return w;
    }
}